### PR TITLE
NEO-2189 final review

### PR DESCRIPTION
**Pull request by yangAtSpoken**
[Link to updated components in Deploy Preview](https://deploy-preview-459--neo-react-library-storybook.netlify.app/?path=/story/components-table--custom-actions)

**Before tagging the team for a review, I have done the following:**

- [x] run `yarn all` locally: ensures that all tests pass, formatting is done, types pass, and builds pass
- [x] reviewed my code changes
- [x] updated the link at the top of this PR (or remove it if not applicable)
- [x] added any important information to this PR (description, images, GIFs, ect.)
- [x] done an accessibility check on my work (tested with Chrome's `axe Dev Tools`, Mac's VoiceOver, etc.)
- [x] tagged `@avaya-dux/dux-design` if any visual changes have occurred
- [x] tagged `@avaya-dux/dux-devs`

`@avaya-dux/dux-design`: 
- review drag and drop functionality; (turn on draggable switch on the page and adjust pagination)
- review if visual is up to design standards

`@avaya-dux/dux-devs`: 
- add tabindex to make draggable row keyboard accessible

## Summary by CodeRabbit

- **New Features**
  - Enhanced drag and drop functionality for table rows with improved logging.
  - Introduced new sorting and data update logic for the table component.
  - Added new `dataSyncOption` and `clearSortByFuncRef` properties to the table's filter context for better filtering and sorting control.
  - Improved mouse hover effects on table header elements for better user interaction.

- **Bug Fixes**
  - Removed the `readOnly` attribute from the table's checkbox component for improved interactivity.

- **Refactor**
  - Optimized state updates and memoization within the table component for better performance.

- **Style**
  - Updated table header classes to include conditional styling based on state for better visual feedback.

### CODE REVIEW

File: src/components/Table/TableComponents/TableHeader.tsx [hunk 1]
```diff
@@ -124,18 +131,31 @@
...
 				)}
 				{shouldHaveCheckboxColumn && (
-					<th className="neo-table-checkbox-th">
+					<th
+						className="neo-table-checkbox-th"
+						onMouseOver={() => setCheckBoxOver(true)}
+						onFocus={() => setCheckBoxOver(false)}
+						onBlur={() => setCheckBoxOver(false)}
+						onMouseOut={() => setCheckBoxOver(false)}
+					>
```
[the remark this answers is not in the document]
Comment (yangAtSpoken) [reply src/components/Table/TableComponents/TableHeader.tsx]: not this one

File: src/components/Table/DraggableTableRow.tsx [hunk 2]
```diff
@@ -52,6 +52,7 @@ export const DraggableTableRow = <T extends Record<string, any>>({
 				preparedRowProps.className,
 				row.original.disabled && "disabled",
 			)}
+			tabIndex={0}
```
Comment (joe-s-avaya) [inline src/components/Table/DraggableTableRow.tsx:55]: why is this necessary? it is a a11y thing? the dnd was working perfectly from what I say in previous PRs

**Review by joe-s-avaya — APPROVED** [review]
I'm a bit confused as to why these changes are necessary. Is it all just for keyboard usability? But everything is working so I'll approve.

File: src/components/Table/DraggableTableRow.tsx (hunk 2 above)
Comment (yangAtSpoken) [reply src/components/Table/DraggableTableRow.tsx:55]: Without this, user can not navigate through rows using tabs.  A row, with columns of text, without a checkbox,  with the drag handle hidden, can not get keyboard focus without this.

**Comment by Justin-Avaya** [comment]
Dragging and dropping a sorted row, yielded interesting results. After sorting a column (A-Z), DND'ing a sorted row didn't behave as expected. I expected the A-Z filter to be cleared but expected the previously sorted rows to remain in-tact, with any modification I previously made. 

For example:

Action: Set column header sorting to (A-Z)
Output:

- Column header - Sort (A-Z)
- A row
- B row
- C row
- 1 row
- 2 row
- 3 row

Action: I drag "2 row" between "B row" and "C row"
Output:

- Column header - Sort (cleared/none)
- A row
- B row
- 2 row
- C row
- 1 row
- 3 row

As of now, when I drop a row somewhere, I have no idea where it went. See example in link below.

This comment is possibly out of scope for this PR but being that it happened while DND'ing, I felt it was best to at least mention it. 

Will let Dev team decide. 

Example:
https://github.com/user-attachments/assets/9e931127-a4f5-41c0-b593-d8774bade4e1

<hr>

This is just a note to whom it may concern, the drop target (light blue background ui element when in mid-drag gesture) against the white table background does not have enough contrast to pass AA accessibility.

Example:
Screenshot 2024-07-17 at 10 44 18 AM

<hr>

While dragging a row with the multiple switch option engaged, both the formatting (column position and text styles) of the row being dragged gets wonky and the color of the check box turns a different color.

Example:
https://github.com/user-attachments/assets/688879dd-3738-4da2-84c1-430f9674ffbf

**Comment by Justin-Avaya** [comment]
@yangAtSpoken 

Matt and I have met and discussed my last PR comment regarding dragging and dropping an item after a previous sort had occurred. 

We are in agreement that the expectation is as follows:

Action: Set column header sorting to (A-Z)
Output:

Column header - Sort (A-Z)
A row
B row
C row
1 row
2 row
3 row
Action: I drag "2 row" between "B row" and "C row"
Output:

Column header - Sort (cleared/none)
A row
B row
2 row
C row
1 row
3 row

**Comment by yangAtSpoken** [comment]
@Justin-Avaya @matthewAvaya Please look again.  Addressed all your comments.

**Review by Justin-Avaya — APPROVED** [review]
@yangAtSpoken - Outstanding work. Nicely done.

- [x] - review drag and drop functionality
- [x] - review if visual is up to design standards

✅ Approved

**Comment by yangAtSpoken** [comment]
@matthewAvaya Needs your approval, thx.

**Review by matthewAvaya — APPROVED** [review]

**Comment by matthewAvaya** [comment]
When I hover over the main header checkbox, first it hovers a small strip on the left side before the entire background turns to the hover state. See video.

https://github.com/user-attachments/assets/190220f3-4cf7-4534-80e5-fd795b024312

**Pull request merged**